File: python/scenario/connected_agent.py
"""
A connected agent function as a scenario agent.

``langwatch.connect_agent`` returns an object that is still the original
function, and that also answers ``invoke(call)`` with the turn fields of the
connected agent contract: ``messages``, ``new_messages``, ``thread_id``,
``session``, ``trace_id`` and the run ``parameters``. This module wraps such
an object into an :class:`AgentAdapter`, so ``scenario.run(agents=[...])``
accepts it directly and the local test and the platform run share one piece
of code for the agent.

The accept is duck typed. Nothing here imports the LangWatch SDK: any callable
with a ``name`` and an awaitable ``invoke`` is a connected agent.

See ``specs/connected-agent-adapter.feature``.
"""

from dataclasses import dataclass, field
from typing import (
    Any,
    Awaitable,
    Dict,
    List,
    Mapping,
    Optional,
    Protocol,
    Sequence,
    Tuple,
    TypeGuard,
    Union,
    cast,
    runtime_checkable,
)

from .agent_adapter import AgentAdapter
from .types import AgentInput, AgentReturnTypes, AgentRole


@dataclass
class ConnectedAgentCall:
    """One turn as the connected agent contract sends it.

    The attribute names match the ``AgentCall`` of the LangWatch SDK, so the
    decorated function reads the same fields locally and on the platform.
    """

    messages: List[Any]
    new_messages: List[Any] = field(default_factory=list)
    thread_id: str = ""
    session: Any = None
    trace_id: Optional[str] = None
    parameters: Dict[str, Any] = field(default_factory=dict)

    def turn_fields(self) -> Dict[str, Any]:
        return {
            "messages": self.messages,
            "new_messages": self.new_messages,
            "thread_id": self.thread_id,
            "session": self.session,
            "trace_id": self.trace_id,
        }


@runtime_checkable
class ConnectedAgentLike(Protocol):
    """The shape ``langwatch.connect_agent`` returns.

    ``invoke`` receives a :class:`ConnectedAgentCall` and returns the reply:
    a string, one message, a list of messages, or an object with ``output``
    and ``session``.
    """

    name: str

    def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    def invoke(self, call: Any) -> Awaitable[Any]: ...


AgentLike = Union[AgentAdapter, ConnectedAgentLike]


def is_connected_agent(value: Any) -> TypeGuard[ConnectedAgentLike]:
    """True for a callable with a string ``name`` and a callable ``invoke``."""
    return (
        not isinstance(value, AgentAdapter)
        and callable(value)
        and isinstance(getattr(value, "name", None), str)
        and callable(getattr(value, "invoke", None))
    )


def trace_id_from_headers(headers: Mapping[str, str]) -> Optional[str]:
    """The trace id of a W3C ``traceparent`` header, or None."""
    traceparent = headers.get("traceparent")
    if not isinstance(traceparent, str):
        return None
    parts = traceparent.split("-")
    if len(parts) >= 3 and len(parts[1]) == 32:
        return parts[1]
    return None


def read_reply(reply: Any) -> Tuple[AgentReturnTypes, Any]:
    """The output and the session of one reply.

    Accepts a string, one message, a list of messages, a mapping with an
    ``output`` key, or an object with an ``output`` attribute (the SDK's
    ``AgentReply``). The session is None when the reply carries none.
    """
    session: Any = None
    output: Any = reply
    if isinstance(reply, Mapping) and "output" in reply:
        output = reply["output"]
        session = reply.get("session")
    elif not isinstance(reply, (str, list, tuple, Mapping)) and hasattr(
        reply, "output"
    ):
        output = getattr(reply, "output")
        session = getattr(reply, "session", None)

    if output is None:
        output = ""
    elif isinstance(output, tuple):
        output = list(output)
    elif not isinstance(output, (str, list, Mapping)):
        output = str(output)
    return cast(AgentReturnTypes, output), session


class ConnectedAgentAdapter(AgentAdapter):
    """Runs a connected agent function as the agent under test.

    Builds the connected call from the scenario's :class:`AgentInput`, keeps
    the ``session`` the function returns per thread, and sends it back on the
    next turn of the same thread, the same echo the platform does.

    Args:
        agent: The decorated function.
        parameters: The run parameters for every call. A parameter that is
            not set here takes the default the function declares.
    """

    role = AgentRole.AGENT

    def __init__(
        self,
        agent: ConnectedAgentLike,
        parameters: Optional[Mapping[str, Any]] = None,
    ) -> None:
        if not is_connected_agent(agent):
            raise TypeError(
                "ConnectedAgentAdapter expects the object langwatch.connect_agent "
                "returns: a callable with a name and an invoke method"
            )
        self.agent = agent
        self.name = agent.name
        self.parameters: Dict[str, Any] = dict(parameters or {})
        self._sessions: Dict[str, Any] = {}

    def session_for(self, thread_id: str) -> Any:
        """The session held for a thread, None before the first reply."""
        return self._sessions.get(thread_id)

    async def call(self, input: AgentInput) -> AgentReturnTypes:
        call = ConnectedAgentCall(
            messages=list(input.messages),
            new_messages=list(input.new_messages),
            thread_id=input.thread_id,
            session=self._sessions.get(input.thread_id),
            trace_id=trace_id_from_headers(input.propagation_headers),
            parameters=dict(self.parameters),
        )
        reply = await self.agent.invoke(call)
        output, session = read_reply(reply)
        if session is not None:
            self._sessions[input.thread_id] = session
        return output


def resolve_agents(
    agents: Sequence[AgentLike],
    parameters: Optional[Mapping[str, Any]] = None,
) -> List[AgentAdapter]:
    """Every connected agent function wrapped, every adapter as is."""
    return [
        ConnectedAgentAdapter(agent, parameters)
        if is_connected_agent(agent)
        else cast(AgentAdapter, agent)
        for agent in agents
    ]

File: javascript/src/agents/connected-agent.ts
/**
 * A connected agent function as a scenario agent.
 *
 * `connectAgent` from the LangWatch SDK returns a function that is still
 * directly callable with the turn fields of the connected agent contract:
 * `messages`, `newMessages`, `threadId`, `session`, `params` and `traceId`.
 * This module wraps it into an {@link AgentAdapter}, so
 * `scenario.run({ agents: [...] })` accepts it directly and the local test
 * and the platform run share one piece of code for the agent.
 *
 * @see specs/connected-agent-adapter.feature
 */
import type { ModelMessage } from "ai";
import {
  AgentAdapter,
  AgentRole,
  isConnectedAgent,
  type AgentInput,
  type AgentReturnTypes,
  type ConnectedAgentCall,
  type ConnectedAgentFunction,
  type ConnectedAgentParameterValue,
} from "../domain";

/** The trace id of a W3C `traceparent` header, empty when there is none. */
export function traceIdFromHeaders(headers: Record<string, string> | undefined): string {
  const traceparent = headers?.traceparent;
  if (typeof traceparent !== "string") return "";
  const parts = traceparent.split("-");
  return parts.length >= 3 && parts[1]?.length === 32 ? parts[1] : "";
}

const isMessage = (value: unknown): value is ModelMessage =>
  typeof value === "object" &&
  value !== null &&
  !Array.isArray(value) &&
  typeof (value as { role?: unknown }).role === "string";

/**
 * The output and the session of one reply. Accepts a string, one message, a
 * list of messages, or `{ output, session }`. The session is `undefined`
 * when the reply carries none.
 */
export function readReply(reply: unknown): { output: AgentReturnTypes; session: unknown } {
  let output: unknown = reply;
  let session: unknown = undefined;
  if (typeof reply === "object" && reply !== null && !Array.isArray(reply) && "output" in reply) {
    const wrapped = reply as { output: unknown; session?: unknown };
    output = wrapped.output;
    session = wrapped.session;
  }
  if (output === null || output === undefined) return { output: "", session };
  if (typeof output === "string") return { output, session };
  if (Array.isArray(output)) return { output: output.filter(isMessage), session };
  if (isMessage(output)) return { output, session };
  return { output: String(output), session };
}

/**
 * Runs a connected agent function as the agent under test.
 *
 * Builds the connected call from the scenario's {@link AgentInput}, keeps the
 * `session` the function returns per thread, and sends it back on the next
 * turn of the same thread, the same echo the platform does.
 */
export class ConnectedAgentAdapter implements AgentAdapter {
  readonly name: string;
  readonly role = AgentRole.AGENT;
  readonly parameters: Record<string, ConnectedAgentParameterValue>;
  private readonly sessions = new Map<string, unknown>();

  constructor(
    readonly agent: ConnectedAgentFunction,
    parameters?: Record<string, ConnectedAgentParameterValue>,
  ) {
    if (!isConnectedAgent(agent)) {
      throw new Error(
        "ConnectedAgentAdapter expects the function connectAgent returns: a function with a name and an environment",
      );
    }
    this.name = agent.name;
    this.parameters = { ...parameters };
  }

  /** The session held for a thread, `undefined` before the first reply. */
  sessionFor(threadId: string): unknown {
    return this.sessions.get(threadId);
  }

  async call(input: AgentInput): Promise<AgentReturnTypes> {
    const call: ConnectedAgentCall = {
      messages: [...input.messages],
      newMessages: [...input.newMessages],
      threadId: input.threadId,
      session: this.sessions.get(input.threadId) ?? null,
      params: { ...this.parameters },
      traceId: traceIdFromHeaders(input.propagationHeaders),
    };
    const invoke = this.agent as unknown as (call: ConnectedAgentCall) => unknown;
    const { output, session } = readReply(await invoke(call));
    if (session !== undefined) {
      this.sessions.set(input.threadId, session);
    }
    return output;
  }
}

/** Every connected agent function wrapped, every adapter as is. */
export function resolveAgents(
  agents: ReadonlyArray<AgentAdapter | ConnectedAgentFunction>,
  parameters?: Record<string, ConnectedAgentParameterValue>,
): AgentAdapter[] {
  return agents.map((agent) =>
    isConnectedAgent(agent) ? new ConnectedAgentAdapter(agent, parameters) : agent,
  );
}
